cmd/amp/cli/cli_test.go

```diff
@@ -1,134 +1,116 @@
 package cli_test
 
 import (
+	"fmt"
 	"github.com/appcelerator/amp/api/server"
 	"gopkg.in/yaml.v2"
 	"io/ioutil"
+	"os"
 	"os/exec"
 	"path"
 	"regexp"
 	"strings"
 	"testing"
-	"time"
 )
 
 type TestSpec struct {
-	fileName string
-	contents []byte
-	valid    bool
+	Name     string
+	Commands []CommandSpec
 }
 
 type CommandSpec struct {
-	Cmd         string   `yaml:"cmd"`
-	Args        []string `yaml:"args"`
-	Options     []string `yaml:"options"`
-	Expectation string   `yaml:"expectation"`
+	Cmd               string   `yaml:"cmd"`
+	Args              []string `yaml:"args"`
+	Options           []string `yaml:"options"`
+	Expectation       string   `yaml:"expectation"`
+	ExpectErrorStatus bool     `yaml:"expectErrorStatus"`
 }
 
 var (
 	testDir = "./test_samples"
 )
 
-func TestMain(t *testing.T) {
-	_, conn := server.StartTestServer()
-	t.Log(conn)
+func TestMain(m *testing.M) {
+	server.StartTestServer()
+	os.Exit(m.Run())
 }
 
 func TestCmds(t *testing.T) {
-	tests := loadFiles(t)
+	tests, err := loadTestSpecs()
+	if err != nil {
+		t.Errorf("unable to load test specs, reason: %v", err)
+		return
+	}
+
 	for _, test := range tests {
 		t.Log("-----------------------------------------------------------------------------------------")
-		t.Logf("test %s\n", test.fileName)
-		parseCmd(t, test)
+		t.Logf("Running spec: %s", test.Name)
+		if err := runTestSpec(t, test); err != nil {
+			t.Error(err)
+			return
+		}
 	}
 }
 
-func loadFiles(t *testing.T) []*TestSpec {
-	tests := []*TestSpec{}
+func loadTestSpecs() ([]*TestSpec, error) {
 	files, err := ioutil.ReadDir(testDir)
 	if err != nil {
-		t.Error(err)
-		return nil
+		return nil, err
 	}
-	for _, f := range files {
-		name := f.Name()
-		t.Log("Loading file:", name)
-		valid := false
-		if !strings.HasPrefix(name, "00-") {
-			valid = true
-		}
-		contents, err := ioutil.ReadFile(path.Join(testDir, name))
+
+	tests := []*TestSpec{}
+	for _, file := range files {
+		test, err := loadTestSpec(path.Join(testDir, file.Name()))
 		if err != nil {
-			t.Errorf("unable to load test sample: %s. Error: %v", name, err)
-		}
-		testSpec := &TestSpec{
-			fileName: name,
-			contents: contents,
-			valid:    valid,
+			return nil, err
 		}
-		tests = append(tests, testSpec)
+		tests = append(tests, test)
 	}
-	return tests
+	return tests, nil
 }
 
-func parseCmd(t *testing.T, test *TestSpec) {
-	commandMap, err := generateCmdSpec(test.contents)
+func loadTestSpec(fileName string) (*TestSpec, error) {
+	content, err := ioutil.ReadFile(fileName)
 	if err != nil {
-		t.Error(err)
-		return
+		return nil, fmt.Errorf("unable to load test spec: %s. Error: %v", fileName, err)
 	}
-	for _, cmdSpec := range commandMap {
-		cmdString := generateCmdString(cmdSpec)
-		t.Log(cmdString, "Command passed.")
-		for i := 0; i < 10; i++ {
```

Comment:
Originally used a small loop to iterate over the command 10 times. The service create command does not block until the service is built in the swarm. 

Tests are currently failing due to this.
`
--- FAIL: TestCmds (0.22s)
        cli_test.go:46: -----------------------------------------------------------------------------------------
        cli_test.go:47: Running spec: test_samples/logs-all.yml
        cli_test.go:97: Running: amp logs
        cli_test.go:46: -----------------------------------------------------------------------------------------
        cli_test.go:47: Running spec: test_samples/logs-metadata.yml
        cli_test.go:97: Running: amp logs  -m
        cli_test.go:46: -----------------------------------------------------------------------------------------
        cli_test.go:47: Running spec: test_samples/logs-numbered.yml
        cli_test.go:97: Running: amp logs  -n 10
        cli_test.go:46: -----------------------------------------------------------------------------------------
        cli_test.go:47: Running spec: test_samples/logs-stack.yml
        cli_test.go:97: Running: amp logs  --stack stack1
        cli_test.go:46: -----------------------------------------------------------------------------------------
        cli_test.go:47: Running spec: test_samples/service-create.yml
        cli_test.go:97: Running: amp service create appcelerator/pinger --name pinger -p www:90:3000
        cli_test.go:46: -----------------------------------------------------------------------------------------
        cli_test.go:47: Running spec: test_samples/service-curl.yml
        cli_test.go:97: Running: curl localhost:90/ping
        cli_test.go:49: miss matched expected output:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (7) Failed to connect to localhost port 90: Connection refused

FAIL
exit status 1
FAIL    github.com/appcelerator/amp/cmd/amp/cli 0.813s `